### PR TITLE
[SPARK-14948][SQL] disambiguate attributes in join condition

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

`Dataset.col/apply` returns a column reference, which is pretty useful to deal with duplicated names in join. e.g.
```
val df1 = ... // [a: int, b: int]
val df2 = ...// [b: int, c: int]

df1.join(df2, df1("b") === df2("b"))
df1.join(df2).drop(df2("b"))
...
```

However, this is problematic for self-join, or joining DataFrames derived from the same DataFrame. The reason is that, the column reference returned by `Dataset.col` is actually `AttributeReference`, which means different DataFrames may return same column reference. After join, the right side would be de-duplicated if it has conflicting attributes with the left side, and the column reference returned by right side would be missing after join, or be wrong and refers to columns from the left side. A simple example to illustrate this problem:
```
val left = spark.range(5)
val right = left.filter($"id" < 1)
left.join(right, left("id") === right("id") + 1, "cross") // This returns empty result, which is wrong
```

To fix this issue entirely, we may need to define a real column reference that is globally uique, and design a dataframe lineage mechanism so that we can use column reference from another dataframe in a dataframe operation, e.g.
```
val df3 = df1.join(df2)
df3.drop(df2("b"))
```

This is a lot of work and is too late for 2.3, here I propose a simple and safe solution to disambiguate attributes in join condition only, which is the most common problematic case.

The idea is simple, we assign a globally unique id to each dataframe, via `AnalysisBarrier`. 
`AttributeReference` returned by `Dataset.col` should be associated with the dataframe it comes from, via the barrier id. This barrier id is mostly ignored, and only used when we are de-duplicating the join right side plan, these attributes with barrier id inside join condition would be replaced by the new attributes generated by the right side plan.

## How was this patch tested?

new regression test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin @gatorsmile @viirya @sameeragarwal

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/joins/InnerJoinSuite.scala [hunk 1]
```diff
@@ -76,11 +76,11 @@ class InnerJoinSuite extends SparkPlanTest with SharedSQLContext {
       testName: String,
       leftRows: => DataFrame,
       rightRows: => DataFrame,
-      condition: () => Expression,
+      condition: => Expression,
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/joins/InnerJoinSuite.scala]: not related, but make the code style consistent with other 2 join test suites.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -1234,11 +1234,24 @@ class Dataset[T] private[sql](
...
       }
   }
 
+  private def createCol(name: String): Column = {
+    val expr = resolve(name) transform {
+      case a: AttributeReference =>
+        // Associate the returned `AttributeReference` with the `AnalysisBarrier` of this Dataset,
+        // by putting the barrier id into `AttributeReference.metadata`. This information is only
+        // used to disambiguate the attributes in join condition when resolving self-join and
+        // de-duplicating the right side plan.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Shall we clarify that this metadata will be removed after analysis?

**Comment by viirya** [comment]
Is this only going to fix the case of joining DataFrames derived from the same DataFrame but not for self-joining?

**Comment by cloud-fan** [comment]
> Is this only going to fix the case of joining DataFrames derived from the same DataFrame but not for self-joining?

Yes. I think self-join is not fixable, `df.join(df, df("id") === (df("id") + 1))`, we have no idea what the join condition means. Maybe we should throw exception for this case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -318,7 +318,10 @@ class Analyzer(
         gid: Expression): Expression = {
       expr transform {
         case e: GroupingID =>
-          if (e.groupByExprs.isEmpty || e.groupByExprs == groupByExprs) {
+          def sameExpressions(e1: Seq[Expression], e2: Seq[Expression]): Boolean = {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Is this a bug not related to this PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -730,12 +733,28 @@ class Analyzer(
...
-              case a: Attribute =>
-                dedupAttr(a, attributeRewrites)
+              case a: AttributeReference =>
+                // Only replace `AttributeReference` when the de-duplicated relation is not a
+                // `AnalysisBarrier`, or this `AttributeReference` is not associated with any
+                // `AnalysisBarrier`, or this `AttributeReference` refers to the de-duplicated
+                // `AnalysisBarrier`, i.e. barrierId matches.
+                if (barrierId.isEmpty || !a.metadata.contains(AnalysisBarrier.metadataKey) ||
+                  barrierId.get == a.metadata.getLong(AnalysisBarrier.metadataKey)) {
+                  dedupAttr(a, attributeRewrites)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Looks like it is the same as:
```scala
// When we de-duplicated an `AnalysisBarrier` and this `AttributeReference` is associated with other
// `AnalysisBarrier` different than the de-duplicated one, we don't replace it.

val notToReplace = barrierId.map { id =>
  a.metadata.contains(AnalysisBarrier.metadataKey) &&
    id != a.metadata.getLong(AnalysisBarrier.metadataKey)
}.getOrElse(false)

if (notToReplace) {
  a
} else {
  dedupAttr(a, attributeRewrites)
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I think so, basically we should always use `semanticEquals` when matching expressions.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Anyway my PR exposed this bug as now `Dataset.col` returns a slightly different attribute with a metadata.

**Comment by cloud-fan** [comment]
retest this please

**Review by viirya — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 5]
```diff
@@ -902,9 +902,20 @@ case class Deduplicate(
...
+case class AnalysisBarrier(child: LogicalPlan, id: Long) extends LeafNode {
   override protected def innerChildren: Seq[LogicalPlan] = Seq(child)
   override def output: Seq[Attribute] = child.output
   override def isStreaming: Boolean = child.isStreaming
   override def doCanonicalize(): LogicalPlan = child.canonicalized
+  override protected def stringArgs: Iterator[Any] = Iterator(child)
+}
+
+object AnalysisBarrier {
+  private val curId = new java.util.concurrent.atomic.AtomicLong()
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: We need to update the doc of `AnalysisBarrier `

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan . If this PR is still valid, could you resolve the conflicts?

**Comment by ssimeonov** [comment]
@cloud-fan do you expect to resolve conflict + merge at some point?

**Comment by cloud-fan** [comment]
I'm closing it since `AnalysisBarrier` is no longer there. We should revisit the whole self-join problem and fix it in 3.0, with breaking changes if necessary.

**Pull request closed without merging**